---
title: "NGS_analysis_basic_homework"
author: "Zehao Li"
date: "2022-10-17"
output: html_document
---

## A.Demultiplexing
## download1  file
```{r download1, include=FALSE}
library(ShortRead)  
download.file("http://cluster.hpcc.ucr.edu/~tgirke/HTML_Presentations/Manuals/testdata/samplefastq/data.zip", "data.zip")
unzip("data.zip")
fastq <- list.files("data", "*.fastq$"); fastq <- paste("data/", fastq, sep="")
names(fastq) <- paste("flowcell6_lane", 1:length(fastq), sep="_") 

cutoff<-30
library(Biostrings); library(GenomicRanges)
demultiplex <- function(x, barcode, nreads) {
  f <- FastqStreamer(x, nreads) 
  while(length(fq <- yield(f))) {
    for(i in barcode) {
      pattern <- paste("^", i, sep="")
      fqsub <- fq[grepl(pattern, sread(fq))] 
      if(length(fqsub) > 0) {
        fqsub <- trimTails(fqsub, k=2, a=rawToChar(as.raw(cutoff+33)), successive=FALSE)
          writeFastq(fqsub, paste(x, i, sep="_"), mode="a", compress=FALSE)
      }
    }
  }
}
demultiplex(x=fastq[1], barcode=c("TT", "AA", "GG"), nreads=50)
```

## B.Sequence Parsing
## Download GFF from Halobacterium sp 
```{r download2, include=FALSE}
download.file("https://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.gff", "data/AE004437.gff")
```
## Download genome sequence from halobacterium sp 
```{r download3, include=FALSE}
download.file("https://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.fna", "data/AE004437.fna")

```
## Task 1 Extract gene ranges, parse their sequences from genome and translate them into proteins
```{r Extract, include=FALSE}
chr <- readDNAStringSet("data/AE004437.fna")
library(GenomicRanges); library(rtracklayer) 
gff <- import("data/AE004437.gff")
gffgene <- gff[values(gff)[,"type"]=="gene"] 
```
## Task 1 parse their sequences from genome 
```{r parse, include=FALSE}
gene<-DNAStringSet(Views(chr[[1]], IRanges(start(gffgene), end(gffgene))))
```
## Task1 translate them into proteins and export the file 
```{r translate, include=FALSE}
names(gene) <- values(gffgene)[,"locus_tag"]
pos <- values(gffgene[strand(gffgene) == "+"])[,"locus_tag"]
p1 <- translate(gene[names(gene) %in% pos])
names(p1) <- names(gene[names(gene) %in% pos])
neg <- values(gffgene[strand(gffgene) == "-"])[,"locus_tag"]
p2 <- translate(reverseComplement(gene[names(gene) %in% neg]))
names(p2) <- names(gene[names(gene) %in% neg])
writeXStringSet(c(p1, p2), "./data/mypep.fasta")
```
## Task 2  Reduce overlapping genes and parse their sequences from genome
```{r reduce overlapping, including=FALSE}
reduce1<-reduce(gff)
DNAStringSet(Views(chr[[1]], IRanges(start(reduce1), end(reduce1))))
```
## Task 3 Generate intergenic ranges and parse their sequences from genome
```{r generate intergenic ranges, including=FALSE}
intergenic <- gaps(reduce1) 
DNAStringSet(Views(chr[[1]], IRanges(start(intergenic), end(intergenic))))
```
